maths/fibonacci_sequence_recursion_with_memoization.py

```diff
@@ -0,0 +1,36 @@
+# Fibonacci Sequence Using Recursion and Memoization
+
+# Memoization Concept: https://en.wikipedia.org/wiki/Memoization
+# Fibonacci Concept : https://pt.wikipedia.org/wiki/Sequ%C3%AAncia_de_Fibonacci
+
+cache = dict()
+
+
+def fibonacci(n):
+    """
+    Return number of index n in sequence fibonacci
+    >>> [fibonacci(i) for i in range(12)]
+    [0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89]
+    """
+    aux = n
+
+    if aux in cache:
+        return cache[aux]
+
+    if (n == 0) or (n == 1):
```

Comment:
```suggestion
    if n in {0, 1}:
```

`set` are hashed, so fast lookup.
```console
$ python -m timeit -s "n = 1" "(n == 0) or (n == 1)"
10000000 loops, best of 5: 32.5 nsec per loop

$ python -m timeit -s "n = 1" "n in {0, 1}"
10000000 loops, best of 5: 21 nsec per loop

$ python -m timeit -s "n = 1" "n in (0, 1)"
10000000 loops, best of 5: 27.4 nsec per loop
```